projects/web-wallet/rpc-protocol.md

```diff
@@ -0,0 +1,432 @@
+
+If the user accepts the transaction, we will create a `SendTransaction` model (or `SendTransactionWalletService` if on the wallet-service) and execute the send transaction.
+
+Since `JSON-RPC` is designed considering only a single response per request, we will only respond the request when the transaction either failed to be pushed to the fullnode (or the wallet-service) or when it's successfully pushed.
+
+## 3.2 `htr_createToken`
+
+When the request is received, we will validate if the parameters are valid by following the parameter type definitions in the Reference-Level explanation
+
+If the validation fails, we will respond with a `INVALID_PARAMETERS` error code
+
+At this point, we will call `prepareCreateNewToken` and display the assembled transaction for user validation (UX is yet to be discussed).
+
+If the user accepts the transaction, we will, just like the `htr_sendTx` method, create a `SendTransaction` model and execute the send transaction. Otherwise, we will answer with a `REQUEST_REJECTED` error
+
+## 3.3 `htr_getUtxos`
+
+When the request is received, we will validate if the parameters are valid by following the parameter type definitions in the Reference-Level explanation
+
+If the validation fails, we will respond with a `INVALID_PARAMETERS` error code
+
+At this point, we will call the `getUtxos` method to retrieve `utxos` that fulfill the requested filters.
+
+Before returning, we will display a confirmation screen (UX yet to be discussed) asking for user confirmation by displaying the `utxos` that will be returned to the `dApp`.
+
+If the user accepts the transaction, we will reply with the array of `utxos`. Otherwise, we will answer with a `REQUEST_REJECTED` error
+
+#### Future possibilities
+
+We might display "checkboxes" before each utxos so the user can choose exactly which ones he want to display
+
+## 3.4 `htr_signWithAddress`
+
+When the request is received, we will validate if the parameters are valid by following the parameter type definitions in the Reference-Level explanation
+
+If the validation fails, we will respond with a `INVALID_PARAMETERS` error code
+
+At this point, we will display the `message` and the `address` to the user, asking for his confirmation. If he rejects the modal, we will respond with an `INVALID_PARAMETERS` error code.
+
+If he accepts, we will call the `signMessageWithAddress` method from the lib and answer the request
+
+## 3.5 `htr_getBalance`
+
+When the request is received, we will validate if the parameters are valid by following the parameter type definitions in the Reference-Level explanation
+
+If the validation fails, we will respond with a `INVALID_PARAMETERS` error code
+
+At this point, we will display the `message` and the `address` to the user, asking for his confirmation. If he rejects the modal, we will respond with an `INVALID_PARAMETERS` error code.
```

Comment:
Should we allow users to modify the balance value? We would allow only values smaller than or equal to the current balance.

Reply:
What do you mean? Modify the balance in the client before returning?

Reply:
Let's say I have 1,000 HTR in my wallet. I might be comfortable to share that I have a balance of 50 HTR (which is smaller than the actual balance).

Reply:
I like the idea, done in 0a62d03c35a04cd1c9a7540ccbf0954e91f40e23

Reply:
Maybe add that the chosen balance should never be higher than the current balance for a given token.

Reply:
Done 3f97ffc5d150fa0b9e22b0a16df80e5af2bd4378

Reply:
Great! Thanks!